aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -346,9 +357,10 @@ Tensor& sgn_out(const Tensor& self, Tensor& result) {
 Tensor sgn(const Tensor& self) { return unary_op_impl(self, at::sgn_out); }
 Tensor& sgn_(Tensor& self) { return unary_op_impl_(self, at::sgn_out); }
 
-Tensor& sin_out(const Tensor& self, Tensor& result) { return unary_op_impl_float_out(result, self, sin_stub); }
-Tensor sin(const Tensor& self) { return unary_op_impl_float(self, sin_stub); }
-Tensor& sin_(Tensor& self) { return unary_op_impl_(self, at::sin_out); }
+TORCH_IMPL_FUNC(sin_out) (const Tensor& self, const Tensor& result) {
+  sin_stub(device_type(), *this);
+  TORCH_INTERNAL_ASSERT(result.scalar_type() == output().dtype());
```

Comment:
`output` is defined at this point?

Reply:
It's an instance method I believe.

Reply:
TORCH_IMPL_FUNC is 
```c++
#define TORCH_IMPL_FUNC(name) void structured_##name::impl
```

I don't see where `output().dtype()` comes from  :thinking:

Reply:
This becomes `at::meta::structured_sin::impl`

Reply:
Where `structured_sin` is a class.

Reply:
Oh. Makes sense.